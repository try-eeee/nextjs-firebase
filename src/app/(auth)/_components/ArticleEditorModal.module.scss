.form {
  padding: 1rem;
}

.header {
  padding: 1em;
  background-color: #cccccc;
  font-weight: 700;
  display: flex;
  justify-content: space-between;
  align-items: center;

  .closeIcon {
    cursor: pointer;
  }
}

.footer {
  border-top: 1px solid #cccccc;
  padding: 1em;
  display: flex;
  justify-content: flex-end;
}

.viewLabel {
  padding-bottom: 0.5rem;
}

.viewTitle {
  padding-bottom: 1rem;
}

.viewContent {
  padding-bottom: 1rem;
}

.error {
  color: red;
  font-size: 0.875rem;
  display: flex;
  justify-content: flex-end;
}
