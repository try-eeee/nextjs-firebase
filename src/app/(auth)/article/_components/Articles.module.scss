.articlesHeader {
  display: flex;
  justify-items: end;
}

.editCell {
  display: flex;
  align-items: flex-start;

  svg {
    margin: 0.6rem;
  }

  @media (max-width: 768px) {
    justify-content: end;
  }
}
