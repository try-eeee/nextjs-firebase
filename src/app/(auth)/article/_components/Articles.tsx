"use client";

import {
  Table,
  TableCell,
  TableHeader,
  TableRow,
} from "@/components/table/Table";
// import { Button } from "@/components/button/Button";
import type { Article } from "../_utils/asyncFunctions";
import styles from "./Articles.module.scss";
// import { formatDate } from "@/utils/formatDate";
// import { useDialog } from "@/components/dialog/Dialog";
// import { useRef, useState } from "react";
// import { ArticleEditorModal, OperationType } from "./ArticleEditorModal";
// import { FaRegTrashAlt, FaEdit, FaRegEye } from "react-icons/fa";

interface ArticleProps {
  articles: Article[];
  uid: string;
}

export function Article(props: ArticleProps) {
  // const { articles, uid } = props;
  const { articles } = props;

  // const AricleRegisterModal = useRef<HTMLDialogElement>(null);
  // const { showModal } = useDialog(AricleRegisterModal);
  // const [operationType, setOperationType] = useState<OperationType>("create");
  // const [article, setArticle] = useState<Article | undefined>(undefined);

  if (!articles) {
    return <></>;
  }

  // const handleCreateArticle = () => {
  //   setOperationType("create");
  //   setArticle(undefined);
  //   showModal();
  // };

  // const handleDeleteArticle = (selectedArticle: Article) => {
  //   setOperationType("delete");
  //   setArticle(selectedArticle);
  //   showModal();
  // };

  // const handleEditArticle = (selectedArticle: Article) => {
  //   setOperationType("edit");
  //   setArticle(selectedArticle);
  //   showModal();
  // };

  // const handleViewArticle = (selectedArticle: Article) => {
  //   setOperationType("view");
  //   setArticle(selectedArticle);
  //   showModal();
  // };

  return (
    <>
      <div className={styles.articlesHeader}>
        {/* <Button buttonType="default" onClick={handleCreateArticle}>
          新規記事作成
        </Button> */}
      </div>
      <Table
        tableHeader={
          <TableRow>
            <TableHeader>記事ID</TableHeader>
            <TableHeader>更新日時</TableHeader>
            <TableHeader>タイトル</TableHeader>
            <TableHeader>閲覧 / 編集 / 削除</TableHeader>
          </TableRow>
        }
        tableRow={
          <>
            {articles.map((article, rowIndex) => (
              <TableRow key={rowIndex}>
                <TableCell headerLabel="記事ID" width={150}>
                  {article.articleId}
                </TableCell>
                <TableCell headerLabel="更新日時" width={150}>
                  {/* {formatDate(article.createdAt)} */}
                </TableCell>
                <TableCell headerLabel="タイトル">{article.title}</TableCell>
                <TableCell headerLabel="閲覧 / 編集 / 削除" width={180}>
                  {/* <div className={styles.editCell}>
                    <FaRegEye onClick={() => handleViewArticle(article)} />
                    {article.isSelf && (
                      <>
                        <FaEdit onClick={() => handleEditArticle(article)} />
                        <FaRegTrashAlt
                          onClick={() => handleDeleteArticle(article)}
                        />
                      </>
                    )}
                  </div> */}
                </TableCell>
              </TableRow>
            ))}
          </>
        }
      />
      {/* <ArticleEditorModal
        uid={uid}
        article={article}
        operationType={operationType}
        aricleRegisterModal={AricleRegisterModal}
      /> */}
    </>
  );
}
