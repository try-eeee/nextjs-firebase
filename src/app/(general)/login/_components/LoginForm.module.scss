.form {
  display: grid;
  grid-template-columns: 1fr;
  row-gap: 16px;
}

.title {
  font-size: 2rem;
  margin-bottom: 1.5rem;
}

.loginButton {
  margin-top: 14px;
  margin-bottom: 14px;
}

.error {
  color: red;
  font-size: 0.875rem;
  text-align: left;
}
