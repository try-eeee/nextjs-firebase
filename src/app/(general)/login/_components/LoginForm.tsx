"use client";

import { FormEvent, useState } from "react";
import Link from "next/link";
import { useRouter } from "next/navigation";
import { getAuth, signInWithEmailAndPassword } from "firebase/auth";
import { getFirebaseAuth } from "@/firebaseClient";
import { Input } from "@/components/input/Input";
import { Button } from "@/components/button/Button";
import { Label } from "@/components/label/Label";
import styles from "./LoginForm.module.scss";

/**
 * ログインフォーム
 */
export default function LoginForm() {
  const [email, setEmail] = useState("");
  const [password, setPassword] = useState("");
  const [error, setError] = useState("");
  const router = useRouter();

  async function handleSubmit(event: FormEvent) {
    event.preventDefault();
    setError("");
    try {
      const { clientApp } = getFirebaseAuth();
      if (!clientApp) {
        throw new Error("Missing app!");
      }
      const credential = await signInWithEmailAndPassword(
        getAuth(clientApp),
        email,
        password
      );
      const idToken = await credential.user.getIdToken();
      await fetch("/api/login", {
        headers: { Authorization: `Bearer ${idToken}` },
      });
      router.push("/");
    } catch (e) {
      setError("ログイン情報が正しくありません。");
    }
  }

  return (
    <main>
      <h1 className={styles.title}>ログイン</h1>
      <form onSubmit={handleSubmit} className={styles.form} action="#">
        <div>
          <Label htmlFor="email" align="left">
            メールアドレス
          </Label>

          <Input
            type="email"
            name="email"
            value={email}
            onChange={(e) => setEmail(e.target.value)}
            id="email"
            className=""
            placeholder="name@company.com"
            required
          />
        </div>
        <div className={styles.mb5rem}>
          <Label htmlFor="password" align="left">
            パスワード
          </Label>
          <Input
            type="password"
            name="password"
            value={password}
            onChange={(e) => setPassword(e.target.value)}
            id="password"
            placeholder="••••••••"
            className=""
            required
          />
        </div>
        {error && (
          <div className={styles.error} role="alert">
            <span className="">{error}</span>
          </div>
        )}
        <Button type="submit" className={styles.loginButton}>
          ログイン
        </Button>
        <p>
          まだ会員登録がお済みでないですか?
          <br />
          <Link href="/register" className="">
            新規会員登録は こちら
          </Link>
        </p>
      </form>
    </main>
  );
}
