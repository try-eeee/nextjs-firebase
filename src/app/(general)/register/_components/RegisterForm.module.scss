.wrapper {
  padding: 1rem;
  margin: 1rem;
}

.form {
  display: grid;
  grid-template-columns: repeat(2, 1fr);
  gap: 1rem;
  max-width: 800px;
  margin: 0 auto;
}

.formItem {
  display: flex;
  flex-direction: column;
}

.fullWidth {
  grid-column: span 2;
}

.alignCenter {
  display: flex;
  flex-direction: column;
  gap: 0.5rem;
  text-align: center;
}

.alignLeft {
  text-align: left;
}

.title {
  font-size: 2rem;
  margin-bottom: 1.5rem;
}

.error {
  color: red;
  font-size: 0.875rem;
  text-align: left;
}
