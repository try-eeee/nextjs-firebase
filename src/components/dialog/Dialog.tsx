"use client";

import styles from "./Dialog.module.scss";
import { forwardRef, RefObject, useCallback } from "react";

interface DialogProps {
  children: React.ReactNode;
  handleOutside: () => void;
}

export function useDialog(ref: RefObject<HTMLDialogElement>) {
  const showModal = useCallback(() => {
    if (ref.current) {
      ref.current.showModal();
    }
  }, [ref]);

  const closeModal = useCallback(() => {
    if (ref.current) {
      ref.current.close();
    }
  }, [ref]);

  return { ref, showModal, closeModal };
}

export const Dialog = forwardRef<HTMLDialogElement, DialogProps>(
  function DialogBase(props, ref) {
    const { children, handleOutside } = props;
    return (
      <dialog ref={ref} onClick={handleOutside} className={styles.dialog}>
        <div className={styles.contents} onClick={(e) => e.stopPropagation()}>
          {children}
        </div>
      </dialog>
    );
  }
);
