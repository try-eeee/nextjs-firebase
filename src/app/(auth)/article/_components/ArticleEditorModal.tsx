"use client";

import clsx from "clsx";
import { Button } from "@/components/button/Button";
import styles from "./ArticleEditorModal.module.scss";
import { Dialog, useDialog } from "@/components/dialog/Dialog";
import { Textarea } from "@/components/textarea/Textarea";
import { zodResolver } from "@hookform/resolvers/zod";
import { RefObject, useEffect } from "react";
import { useForm } from "react-hook-form";
import { articleEditorSchema } from "../_utils/articleEditorSchema";
import {
  createArticle,
  CreateArticle,
  deleteArticle,
  updateArticle,
} from "../actions";
import { Article } from "../_utils/asyncFunctions";
import { Input } from "@/components/input/Input";
import { Label } from "@/components/label/Label";
import { FirebaseError } from "firebase-admin";
import { IoMdClose } from "react-icons/io";

export type OperationType = "edit" | "create" | "delete" | "view";

interface ArticleEditorModalProps {
  operationType: OperationType;
  aricleRegisterModal: RefObject<HTMLDialogElement>;
  article?: Article;
  uid: string;
}

export function ArticleEditorModal(props: ArticleEditorModalProps) {
  const { operationType, aricleRegisterModal, article, uid } = props;
  const { ref: myRef, closeModal } = useDialog(aricleRegisterModal);

  useEffect(() => {
    reset({
      title: article?.title ?? "",
      content: article?.content ?? "",
    });
  }, [article]);

  const {
    register,
    handleSubmit,
    setValue,
    getValues,
    reset,
    formState: { errors },
  } = useForm<CreateArticle>({
    mode: "onSubmit",
    resolver: zodResolver(articleEditorSchema),
  });

  const onSubmit = async (data: CreateArticle) => {
    try {
      await {
        create: () => createArticle({ ...data, uid }),
        edit: () =>
          updateArticle({
            ...data,
            articleId: article?.articleId ?? "",
            uid,
          }),
        delete: () => {
          return;
        },
        view: () => {
          return;
        },
      }[operationType]();
    } catch (e) {
      const errorCode = (e as FirebaseError).code;
      const errorMessage = (e as FirebaseError).message;
      console.error(`Error Code: ${errorCode}, Message: ${errorMessage}`);
    } finally {
      reset();
      closeModal();
    }
  };

  const handleDelete = async () => {
    if (!article?.articleId) {
      console.error("");
      return;
    }
    try {
      await deleteArticle({
        uid,
        articleId: article.articleId,
      });
    } catch (e) {
      const errorCode = (e as FirebaseError).code;
      const errorMessage = (e as FirebaseError).message;
      console.error(`Error Code: ${errorCode}, Message: ${errorMessage}`);
    } finally {
      reset();
      closeModal();
    }
  };

  const handleCloseModal = () => {
    reset();
    closeModal();
  };

  return (
    <Dialog ref={myRef} handleOutside={handleCloseModal}>
      <>
        <div
          className={clsx(
            styles.header,
            {
              create: styles.create,
              edit: styles.edit,
              delete: styles.delete,
              view: styles.view,
            }[operationType]
          )}
        >
          <span>
            {
              {
                create: "新規作成",
                edit: "編集",
                delete: "削除",
                view: "詳細",
              }[operationType]
            }
          </span>
          <IoMdClose className={styles.closeIcon} onClick={handleCloseModal} />
        </div>
        <div className={styles.form}>
          {operationType === "view" ? (
            <div className={styles.viewLabel}>タイトル</div>
          ) : (
            <Label htmlFor="title" align="left" required>
              タイトル
            </Label>
          )}
          {operationType === "view" ? (
            <div className={styles.viewTitle}>{getValues("title")}</div>
          ) : (
            <Input
              type="text"
              register={register("title", {
                required: true,
              })}
              onBlur={(e) => setValue("title", e.target.value.trim())}
              disabled={operationType === "delete"}
            />
          )}
          {errors.title && (
            <span className={styles.error}>タイトルは必須です</span>
          )}

          {operationType === "view" ? (
            <div className={styles.viewLabel}>内容</div>
          ) : (
            <Label htmlFor="content" align="left" required>
              内容
            </Label>
          )}
          {operationType === "view" ? (
            <div className={styles.viewContent}>{getValues("content")}</div>
          ) : (
            <Textarea
              register={register("content", {
                required: true,
              })}
              maxLength={140}
              disabled={operationType === "delete"}
            />
          )}

          {errors.content && (
            <span className={styles.error}>内容は必須です</span>
          )}
        </div>

        <div className={operationType === "view" ? "" : styles.footer}>
          {
            {
              create: (
                <Button onClick={handleSubmit(onSubmit)}>新規投稿</Button>
              ),
              edit: <Button onClick={handleSubmit(onSubmit)}>編集</Button>,
              delete: (
                <Button buttonType="danger" onClick={handleDelete}>
                  削除
                </Button>
              ),
              view: <></>,
            }[operationType]
          }
        </div>
      </>
    </Dialog>
  );
}
