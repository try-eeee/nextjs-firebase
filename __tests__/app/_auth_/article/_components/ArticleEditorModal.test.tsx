import React, { createRef } from "react";
import { render } from "@testing-library/react";
import { ArticleEditorModal } from "@/app/(auth)/article/_components/ArticleEditorModal";
import { useDialog } from "@/components/dialog/Dialog";

jest.mock("@/app/(auth)/article/actions", () => ({
  createArticle: jest.fn(),
  updateArticle: jest.fn(),
  deleteArticle: jest.fn(),
}));

describe("ArticleEditorModal UnitTest", () => {
  const refArticleEditorModal = createRef<HTMLDialogElement>();

  test("Snapshot Test Of ArticleEditorModal", () => {
    const { asFragment } = render(
      <ArticleEditorModal
        operationType="view"
        aricleRegisterModal={refArticleEditorModal}
        uid="testUid"
      />
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
