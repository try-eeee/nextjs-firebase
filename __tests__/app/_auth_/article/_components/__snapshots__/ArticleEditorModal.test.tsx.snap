// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ArticleEditorModal UnitTest Snapshot Test Of ArticleEditorModal 1`] = `
<DocumentFragment>
  <dialog
    class="dialog"
  >
    <div
      class="contents"
    >
      <div
        class="header view"
      >
        <span>
          詳細
        </span>
        <svg
          class="closeIcon"
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 512 512"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M405 136.798L375.202 107 256 226.202 136.798 107 107 136.798 226.202 256 107 375.202 136.798 405 256 285.798 375.202 405 405 375.202 285.798 256z"
          />
        </svg>
      </div>
      <div
        class="form"
      >
        <div
          class="viewLabel"
        >
          タイトル
        </div>
        <div
          class="viewTitle"
        />
        <div
          class="viewLabel"
        >
          内容
        </div>
        <div
          class="viewContent"
        />
      </div>
      <div
        class=""
      />
    </div>
  </dialog>
</DocumentFragment>
`;
